### PR TITLE
mTLS performance overhead measurement test suite

### PR DESCRIPTION
**Pull request by incfly**
- 12% ish cpu overhead and P99 latency increased. no surprise.
- readme has reproducible setup instructions.
- claim ownership.

### CODE REVIEW

File: perf/security/mtls/templates/service-graph.yaml [hunk 1]
```diff
@@ -0,0 +1,246 @@
+apiVersion: v1
+data:
+  service-graph: |
+    services:
```
Comment (mandarjog) [inline perf/security/mtls/templates/service-graph.yaml]: Since we are just measuring mtls overhead, we don’t need service graph at all.
You should model this on the benchmark test which has fewer moving parts.

File: perf/istio-install/automtls.yaml [hunk 2]
```diff
@@ -3,6 +3,8 @@ kind: IstioControlPlane
 spec:
   values:
     global:
+      sidecarInjectorWebhook:
+        rewriteAppHTTPProbe: true
```
Comment (mandarjog) [inline perf/istio-install/automtls.yaml]: Why is this needed for the mtls overhead test?

Comment (incfly) [reply perf/istio-install/automtls.yaml]: turning on mtls blocks health check. this solves the problem.

does no harm anyway,, only impacting sidecar injector

File: perf/security/mtls/README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+- [Stackoverflow](https://stackoverflow.com/questions/1615882/how-much-network-overhead-does-tls-add-compared-to-a-non-encrypted-connection)
+new TLS session requries 6.5K bytes, the messsage itself is the same size.
+
+## Misc
+
+TODO
+
+- stalled here, need to find some data on where is stored, png files.
+- no traffic when plaintext.
+- probe rewrite does not seem work with tls
```
Comment (mandarjog) [inline perf/security/mtls/README.md]: Please add png files to a gcs bucket instead. This bloats the repository

**Comment by mandarjog** [comment]
@carolynhu @richardwxn what would it take to just add 1 more config that changes destination rule to plaintext to our nightly tests? We do need to run it a bit longer and ensure that cpu is captured correctly

File: perf/istio-install/automtls.yaml (hunk 2 above)
Comment (richardwxn) [reply perf/istio-install/automtls.yaml]: Just to double check, is your automtls.yaml overlays on top of the operator built-in default profile or the operator_default.yaml defined here

File: perf/security/mtls/README.md (hunk 3 above)
Comment (richardwxn) [reply perf/security/mtls/README.md]: fyi we upload the benchmark data to istio-build/perf under istio-testing project

File: perf/security/mtls/setup.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      # Generate larger load for load client.
+      export LOADCLIENT_EXTRA_HELM_FLAGS="--set qps=200"
+      ./setup_test.sh "${NAMESPACE}" "svc-"
+      popd || exit
+  fi
+}
+
+function setup_istio() {
+  pushd ../../istio-install || exit
+  export OPERATOR_PROFILE="automtls.yaml" OPERATOR_SHA='438827b1602037fe30dedcd0008ce0cae0ef0aee' &&
```
Comment (richardwxn) [inline perf/security/mtls/setup.sh]: can you update the operator sha file instead

**Comment by richardwxn** [comment]
@mandarjog @incfly I think the question is whether we run the overhead test with existing benchmark or service graph(actually I have another pending PR to add service graph test to jobs). If running with benchmark job, config change is simple, just add plaintext as well, but we have to run longer time for measuring, now the overall job takes 11.5h.

File: perf/security/mtls/setup.sh (hunk 4 above)
Comment (incfly) [reply perf/security/mtls/setup.sh]: This SHA is grabbed from 1.4 HEAD.

If I remembered correctly, without being specific in the test, the HEAD sha or the sha file is not working for my test suite.

Maybe I'll remove this one and add doc in README instead.

I've experienced several times of test suite is broken after a while, due to the dependency changes, (load client installation), or the Istio installation slight changes.

I think the solution should be having each test suite defines some validation hook, and setup some CI to verify the installation meets the test suites' needs.

```shell
source setup.sh && validate
```

**Comment by incfly** [comment]
From my experience, noticing the data plane cpu and p99 latency difference, 1hr running on each setup is enough to see the difference. So it'll be around 2hrs in total. but as mandar said, to get more confidence, a bit longer will be better.

I'm happy with either way. Haven't learned benchmark test suite setup basics, so might use this temporarily for the automtls rollout. I do feel it's nicer to have the same set of config language, like service graph, to describe the suite. Could it be single a -> b service graph in perf benchmark?

Maybe we could proceed first and see how to migrate to benchmark later?

File: perf/security/mtls/README.md (hunk 3 above)
Comment (incfly) [reply perf/security/mtls/README.md]: done.

File: perf/istio-install/automtls.yaml (hunk 2 above)
Comment (incfly) [reply perf/istio-install/automtls.yaml]: I'm using whatever defined in the `setup_istio_operator.sh`, open to update that shell library.

https://github.com/istio/tools/blob/f9fdb1d49b55df4ca63447aaab16398c68024fa2/perf/istio-install/setup_istio_operator.sh#L63

Propogates as

```
go run ./cmd/mesh.go manifest apply -f "${CR_FILENAME}" --force=true --set defaultNamespace=${defaultNamespace}
```

THe `CR_FILENAME` is replaced, that might be not overlay?

File: perf/security/mtls/setup.sh (hunk 4 above)
Comment (incfly) [reply perf/security/mtls/setup.sh]: moved to instructions instead.

**Comment by mandarjog** [comment]
@incfly thanks for being understanding.
I would like this to be integrated as another config in the two pod tests.
@richardwxn can you work with incfly come up with a plan?
As for the length of test we can run a few times as decide what is needed.
I don’t think a 2 hr test is needed, 5-8 mins seems reasonable.

**Comment by incfly** [comment]
I'm closing this PR, and follow up with @richardwxn later to make it as two pods.
